OverlayPlugin.Core/MemoryProcessors/Combatant/LineCombatant.cs

```diff
@@ -0,0 +1,411 @@
+                    {
+                        var diff = Math.Abs((combatant.Heading + Math.PI) - (oldCombatant.Heading + Math.PI));
+                        var checkDiff = inCombat ? CombatantChangeCriteria.InCombatDistanceHeading : CombatantChangeCriteria.OutOfCombatDistanceHeading;
+                        if (diff > checkDiff)
+                        {
+                            changed.Add(combatant.GetType().GetField("Heading"));
+                        }
+                    }
+                }
+                if (changed.Count > 0)
+                {
+                    combatantStateMap[combatant.ID] = new CombatantStateInfo()
+                    {
+                        lastUpdated = now,
+                        combatant = combatant,
+                    };
+                    WriteLine(
+                        CombatantMemoryChangeType.Change,
+                        combatant.ID,
+                        string.Join("", changed.Select((fi) => FormatFieldChange(fi, combatant))));
+                }
+            }
+
+            // Any combatants no longer in memory, signify that they were removed
+            foreach (var ID in combatantStateMap.Keys)
+            {
+                // If we're filtering, only consider removing those in the filters
+                if (filter.Length > 0 && !filter.Contains(ID))
+                {
+                    continue;
+                }
+                if (!combatants.Any((c) => c.ID == ID))
+                {
+                    combatantStateMap.TryRemove(ID, out var _);
+                    WriteLine(CombatantMemoryChangeType.Remove, ID, "");
+                }
+            }
+        }
+
+        private string FormatFieldChange(FieldInfo info, Combatant combatant, bool skipDefaultValues = false)
+        {
+            var value = info.GetValue(combatant);
+
+            if (value == null)
+            {
+                if (skipDefaultValues)
+                {
+                    return string.Empty;
+                }
+                return $"|{info.Name}|NULL";
+            }
+
+            if (skipDefaultValues && value.Equals(CombatantChangeCriteria.DefaultValues[info.FieldType]))
+            {
+                return string.Empty;
+            }
+
+            if (info.Name == "PCTargetID" || info.Name == "NPCTargetID" || info.Name == "BNpcNameID" || info.Name == "BNpcID" || info.Name == "TargetID" || info.Name == "OwnerID" || info.Name == "CastTargetID")
+            {
+                return $"|{info.Name}|0x{value:X}";
+            }
+
+            if (info.FieldType.IsEnum)
+            {
+                return $"|{info.Name}|{Convert.ChangeType(value, Enum.GetUnderlyingType(info.FieldType))}";
+            }
+            return $"|{info.Name}|{value}";
+        }
+
+        private void WriteLine(CombatantMemoryChangeType type, uint combatantID, string info)
+        {
+            var line = $"{type}|{combatantID:X8}{info}";
```

Comment:
I had commented in a DM that there is an option to do something more akin to what I do, where you consume the line in the background and use that to feed a state object. So then in the actual trigger you'd be able to do something like `data.combatants[0x12345678].WeaponId == 4`. This doesn't cover cases where the change in combatant state *is* the thing you're trying to trigger off of, though. 

I think there's been some prior convos about this, since 38-lines already work like this (and the effects in 21/22 lines sort of do, but they're padded out with zero fields). Is it possible to make a "special" def where you do something like this?
```ts
{
      type: 'Combatant',
      netRegex: { WeaponId: '4', source: 'Omega-F', capture: false },
}
```

Reply:
This isn't the only situation where there's a repeating field in a log line, this is just the first case where the data is actually wanted directly as part of a trigger.

I think that if we want to match against this in cactbot, we'll need to do one of:
- Handle the repeating fields across all log lines properly
- Treat the remainder of the line as plain text, and then match via something like `{ type: 'Change', contents: '.*WeaponId\\|4.*' }`
- Track state as part of PopupText itself and use that state tracking in other triggers

But I think that's a conversation better had over on the cactbot repo.

Reply:
Well, I wanted to have the conversation here in case there were changes that could be made that would make it easier there (and to double check that you thought this would at least have a plausible solution there).

One thing that would make it easier to turn into a (more efficient) regex would be if the field names were sorted.  I'd like to avoid `'.*WeaponId\\|4.*'` if we can.

Reply:
> One thing that would make it easier to turn into a (more efficient) regex would be if the field names were sorted.

I'm not sure how sorting would help here, unless every single field was always output which is going to massively increase the log file size.

Right now, even if the fields are sorted, you could end up with all of the following possible lines:

```
Change|40021F84|Heading|-4.792213E-05|PosX|114.384|PosY|86.26396|PosZ|0|Radius|10
Change|40021F84|Heading|-4.792213E-05|PosX|114.384|PosY|86.26396|PosZ|0|Radius|10|TargetID|E0000000
Change|40021F84|CurrentHP|20760000|Heading|-4.792213E-05|PosX|114.384|PosY|86.26396|PosZ|0|Radius|10|TargetID|E0000000
Change|40021F84|CurrentHP|20760000|TargetID|E0000000
```

In addition, this is going to end up splitting up common values, e.g. if `NPCTargetID` changed above, that would end up between `Heading` and `PosX`.

Reply:
Yeah, definitely.  It doesn't make sense to sort `Change` lines as only certain values are there.  Sorting the full lines makes some sense to me, because then you could write a regex trigger that matched multiple values (which I think would be impossible otherwise?).  But maybe that's not relevant because full lines only get written on infrequently changing values so it's not that useful.

Reply:
Oops sorry, then this is my misunderstanding based on past commits.